test_fakenewsredis.py

```diff
@@ -3046,6 +3046,212 @@ def test_set_existing_key_persists(self):
+        self.assertEqual(val, [b'bar1', b'bar2'])
+
+    def test_eval_mget_none(self):
+        self.redis.set('{foo}1', None)
+        self.redis.set('{foo}2', None)
+        val = self.redis.eval('return redis.call("mget", "{foo}1", "{foo}2")', 2, 'foo1', 'foo2')
+        self.assertEqual(val, [b'None', b'None'])
+
+    def test_eval_mget_not_set(self):
+        val = self.redis.eval('return redis.call("mget", "{foo}1", "{foo}2")', 2, 'foo1', 'foo2')
+        self.assertEqual(val, [None, None])
+
+    def test_eval_hgetall(self):
+        self.redis.hset('foo', 'k1', 'bar')
+        self.redis.hset('foo', 'k2', 'baz')
+        val = self.redis.eval('return redis.call("hgetall", "foo")', 1, 'foo')
+        sorted_val = sorted(
+            [
+                val[:2],
+                val[2:]
+            ]
+        )
+        self.assertEqual(
+            sorted_val,
+            [[b'k1', b'bar'], [b'k2', b'baz']]
+        )
+
+    def test_eval_list_with_nil(self):
+        self.redis.lpush('foo', 'bar')
+        self.redis.lpush('foo', None)
+        self.redis.lpush('foo', 'baz')
+        val = self.redis.eval('return redis.call("lrange", KEYS[1], 0, 2)', 1, 'foo')
+        self.assertEqual(val, [b'baz', b'None', b'bar'])
+
+    def test_eval_invalid_command(self):
+        with self.assertRaises(ResponseError):
+            self.redis.eval(
+                'return redis.call("FOO")',
+                0
+            )
+
+    def test_eval_syntax_error(self):
+        with self.assertRaises(ResponseError):
+            self.redis.eval('return "', 0)
+
+    def test_eval_global_variable(self):
+        # Redis doesn't allow script to define global variables
+        with self.assertRaises(ResponseError):
+            self.redis.eval('a=10', 0)
+
+    def test_eval_global_and_return_ok(self):
+        # Redis doesn't allow script to define global variables
+        with self.assertRaises(ResponseError):
+            self.redis.eval(
+                '''
+                a=10
+                return redis.status_reply("Everything is awesome")
+                ''',
+                0
+            )
+
+    def test_eval_convert_number(self):
+        # Redis forces all Lua numbers to integer
+        val = self.redis.eval('return 3.2', 0)
+        self.assertEqual(val, 3)
+        val = self.redis.eval('return 3.8', 0)
+        self.assertEqual(val, 3)
+        val = self.redis.eval('return -3.8', 0)
+        self.assertEqual(val, -3)
+
+    def test_eval_convert_bool(self):
+        # Redis converts true to 1 and false to nil (which redis-py converts to None)
+        val = self.redis.eval('return false', 0)
+        self.assertIsNone(val)
+        val = self.redis.eval('return true', 0)
+        self.assertEqual(val, 1)
+        self.assertNotIsInstance(val, bool)
+
+    def test_eval_convert_nil_to_false(self):
+        val = self.redis.eval('return ARGV[1] == false', 0, None)
+        self.assertFalse(val)
```

Comment:
I don't think this test is testing what is implied by the name. redis-py converts all arguments through its to_bytes, so ``None`` as an argument becomes the string "None" in the script, rather than false. The following test currently fails (but passes for real redis):
```python
        val = self.redis.eval('return ARGV[1] == "None"', 0, None)
        self.assertTrue(val)
```

Reply:
Huh, something went wrong here. I'll sort it out...

Reply:
Actually Lua scripts require that too, now that I think about it. I guess it doesn't matter for the real Redis tests since they just run against a single node.

Reply:
Braces are gone.

Reply:
Fixed.